src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComInterfaceMarshaller.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.Versioning;
+
+namespace System.Runtime.InteropServices.Marshalling
+{
+    /// <summary>
+    /// COM interface marshaller using a StrategyBasedComWrappers instance
+    /// </summary>
+    /// <remarks>
+    /// This marshaller will always pass the <see cref="CreateObjectFlags.Unwrap"/> flag
+    /// to <see cref="ComWrappers.GetOrCreateObjectForComInstance(IntPtr, CreateObjectFlags)"/>.
+    /// </remarks>
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [CLSCompliant(false)]
+    [CustomMarshaller(typeof(CustomMarshallerAttribute.GenericPlaceholder), MarshalMode.Default, typeof(ComInterfaceMarshaller<>))]
+    public static unsafe class ComInterfaceMarshaller<T>
+    {
+        private static readonly Guid? TargetInterfaceIID = StrategyBasedComWrappers.DefaultIUnknownInterfaceDetailsStrategy.GetIUnknownDerivedDetails(typeof(T).TypeHandle)?.Iid;
+
+        public static void* ConvertToUnmanaged(T? managed)
+        {
+            if (managed == null)
+            {
+                return null;
+            }
+            if (!ComWrappers.TryGetComInstance(managed, out nint unknown))
+            {
+                unknown = StrategyBasedComWrappers.DefaultMarshallingInstance.GetOrCreateComInterfaceForObject(managed, CreateComInterfaceFlags.None);
+            }
+            return CastIUnknownToInterfaceType(unknown);
+        }
+
+        public static T? ConvertToManaged(void* unmanaged)
+        {
+            if (unmanaged == null)
+            {
+                return default;
+            }
+            return (T)StrategyBasedComWrappers.DefaultMarshallingInstance.GetOrCreateObjectForComInstance((nint)unmanaged, CreateObjectFlags.Unwrap);
+        }
+
+        internal static void* CastIUnknownToInterfaceType(nint unknown)
+        {
+            if (TargetInterfaceIID is null)
+            {
+                return (void*)unknown;
```

Comment:
Why does this return `unknown` instead of throw InvalidCastException?

Reply:
If we can't find an interface to marshal as, we will marshal as IUnknown. This matches the existing behavior for UnmanagedType.Interface.

I'll add a comment and update the compatibility doc.